### ISSUE TITLE
Champion "Constant interpolated strings"

### ISSUE THREAD

**Issue by agocke** [dotnet/csharplang#2951]

**Comment by Joe4evr** [comment]
![](https://media1.tenor.com/images/05fae48a148f54be961e484d3d582091/tenor.gif?itemid=3718772)

**Comment by svick** [comment]
As I understand it, the biggest problem with this feature is that the value of an interpolated string can depend on the current culture. How do you propose to solve that? In other words, if I had `const string pi = $"{3.14}";`, what would be the value of `pi`?

**Comment by YairHalberstadt** [comment]
@svick 
Ignore that and interpret it as `"" + 3.14`?

**Comment by HaloFour** [comment]
From the [notes](https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-11-11.md):

> A more useful version would allow constant interpolated strings if all of the substitutions are constant strings.

**Comment by ronnygunawan** [comment]
@YairHalberstadt Tried it on C# interactive:
```cs
> "" + 3.14
"3.14"
> System.Threading.Thread.CurrentThread.CurrentCulture =
. System.Globalization.CultureInfo.GetCultureInfo("id-ID");
. "" + 3.14
"3,14"
```

**Comment by 333fred** [comment]
> As I understand it, the biggest problem with this feature is that the value of an interpolated string can depend on the current culture. How do you propose to solve that? In other words, if I had const string pi = $"{3.14}";, what would be the value of pi?

@svick, we haven't come up with a full proposal yet, but in LDM yesterday we were of the opinion that you can't do that. You can only do things in `const` interpolated strings that are themselves `const`. `"" + 3.14` is not a constant expression, so you can't interpolate it.

**Comment by MgSam** [comment]
What are the use cases of this feature?

This feature seems of such minor use that I have no idea how it made it to an LDM.

**Comment by CyrusNajmabadi** [comment]
> What are the use cases of this feature?

Building a constant string out of other constant, but allowing the user to use interpolations, instead of forcing them to use concatenation.

**Comment by HaloFour** [comment]
It's a very slight difference, and the end result is the same.

I was going to mention reusable URI components in routing attributes as a use case, but then I remembered that those attributes use curly braces to denote path parameters, and that could end up being _really_ confusing.  That might be an argument against this feature.

**Comment by CyrusNajmabadi** [comment]
IDE colorization should help with that @HaloFour.  Right?

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> IDE colorization should help with that @HaloFour. Right?

Unless you also have an IDE extension to colorize URIs in ASP.NET routing attributes?  😄

**Comment by MgSam** [comment]
> > What are the use cases of this feature?
> 
> Building a constant string out of other constant, but allowing the user to use interpolations, instead of forcing them to use concatenation.

That's saying how it works, not what the use cases are. 

I can probably count on one hand how many times I could've used a feature like this in the 10+ years I've been writing C#. I don't understand how such a minor feature that no one is asking for makes it to an LDM but related (and much more useful requests), like expanding the set of compile-time allowed expressions don't. 

What about using System.Math at compile time? What about reflection at compile time? All things people ask for ad-nauseam and that solve real problems. Why should string interpolation, which is just a fancy syntax sugar for stuff we can already do with concatenation at compile time, be the blessed operation that gets promoted?

I'm not saying that any of those other things should be prioritized or not prioritized at this time either, it just continues to strike me how completely arbitrary the LDM prioritization continues to be.

**Comment by CyrusNajmabadi** [comment]
> That's saying how it works, not what the use cases are.

The use case is when i want to make a constant string and build it out of other constant strings.  Formattable strings looks nicer and easier to read, so i would like to use those.

>  I don't understand

The LDM decides what they want to champion.  It's that simple.

> What about using System.Math at compile time?

What about it? 

> What about reflection at compile time? 

What about it?

> All things people ask for ad-nauseam and that solve real problems.

Ok.  Convince an LDM member to take that up.  It's not relevant in this proposal.

>  Why should string interpolation, which is just a fancy syntax sugar for stuff we can already do with concatenation at compile time, be the blessed operation that gets promoted?

**Comment by drewnoakes** [comment]
I suspect I would use this most in `DebuggerDisplay` attributes:

```c#
// before
[DebuggerDisplay("{Foo}")]

// or maybe
[DebuggerDisplay("{" + nameof(Foo) + "}")]

// after
[DebuggerDisplay($"{{{nameof(Foo)}}}")]
```

**Comment by drewnoakes** [comment]
Another use case:

```c#
[Obsolete($"Use {nameof(ILaunchSettingsProvider2)} instead.")]
```

**Comment by agocke** [comment]
Yeah, attribute arguments were the place I thought this most useful, and have run into it before. I championed it because I happen to know that we implemented an optimization that basically turns an interpolated string into a string.Concat call not too long ago, so I bet this would be pretty easy to implement.

**Comment by yaakov-h** [comment]
@agocke the latest compiler doesn't just turn it into `string.Concat`. If all pieces are constants, it actually emits the concatenated/interpolated string as a single string:

https://sharplab.io/#v2:D4AQzABCBMEJIBkCGBXAdgYwBYGUCmALgQJZoDmAzgAoBOA9gG7EAmeNsA3gLABQE/EXgKgBGAGwQaeJMzpoANgE9RABggBZPBQpIyeCAF4IAEgBEAVQr6OaJAFs8dAGYAKRKky5CJctXpNWdgBKAF8IUgoCaWYAOlMAbl4QoA==

**Comment by agocke** [comment]
Yup, that's why I thought this would be pretty simple. Probably some stuff to mop up around constant folding, but overall shouldn't be too bad.

**Comment by gafter** [comment]
FYI: `"" + 3.14` is not a constant in C#.

**Comment by uygary** [comment]
@svick has a very fair and extremely important point. Culture dependency and unavoidable side effects that will be introduced by that is making me very wary about this.
Even though I love the uses cases, I'm not sure if this is a good idea.

**Comment by HaloFour** [comment]
@uygary 

This feature won't work in this scenarios.  It'll only work for interpolating other constant strings, so locale is not a concern.  It's really just a slightly different syntax for string concatenation.

**Comment by uygary** [comment]
@HaloFour in that case I retract my objection. Thanks for clearing that up, and sorry for missing that quote from the notes.

**Comment by AustinWise** [comment]
Just for reference, this is a long-standing request. dotnet/roslyn#4678 was migrated from an internal Microsoft bug database in 2015. There are also a few issues in this repo related to this feature:

* #384
* #2077

I look forward to this simple and straightforward improvement to the ergonomics of string interpolation in C#.

**Comment by GlassGrass** [comment]
Can #3044 give benefits this proposal?
When InvariantCulture is specified, not only const string but also const float seems to be available as an element of constant interpolation string.

**Comment by CleanCodeX** [comment]
I would like to see any progress for this. Which are the remaining problems preventing this useful stuff to be implemented?

**Comment by 333fred** [comment]
Nothing. It's implemented in a feature branch and will very likely be in C# 10.

**Comment by 333fred** [comment]
Note: this technically _does_ need an implementation. But we have one, it just needs to be merged (hopefully in the next week or two), at which point we'll move this to the 10.0 milestone.

**Comment by 333fred** [comment]
This has now been merged, and will be available in VS 16.9 under the `preview` language version.

**Comment by CleanCodeX** [comment]
when is 16.9. about to be rolled out?

**Comment by 333fred** [comment]
> when is 16.9. about to be rolled out?

There's no public date I can share currently.

**Comment by jaredpar** [comment]
The best place to look for VS release dates, RTM and preview, is here 

https://docs.microsoft.com/en-us/visualstudio/releases/2019/release-notes

**Comment by CleanCodeX** [comment]
of course, this page I am checking daily. Is there any site where release dates are listed in advance? It's nasty that I daily have to check if there is a new version for VS installer to update. Is there any possibility to get notified via email if a new version is available? This is annoying.

**Comment by 333fred** [comment]
> It's nasty that I daily have to check if there is a new version for VS installer to update.

VS itself does update checks, and should let you know with a popup inside VS if there is a new version availablel.

**Comment by CleanCodeX** [comment]
> VS itself does update checks, and should let you know with a popup inside VS if there is a new version available.

well, sometimes this actually happens... in less than 10% of all cases I downloaded  the new version by myself before VS had any clue of it. not sure what the notification policy behind this is.

**Comment by theunrepentantgeek** [comment]
Based on my own observations, I believe VS checks daily.

**Comment by ViIvanov** [comment]
Is it possible to allow alignment component in interpolation expression?
```cs
const string TitleHeader = "Title";
const string CountHeader = "Count";
const string TableHeader = $"{TitleHeader,-50}|{CountHeader, 10}";
```

**Comment by 333fred** [comment]
No, at least not as proposed and implemented.

**Comment by ViIvanov** [comment]
I think (maybe I'm wrong?) alignment can be implemented for `const` strings. Should I create a proposal for it?

**Comment by 333fred** [comment]
> I think (maybe I'm wrong?) alignment can be implemented for `const` strings. Should I create a proposal for it?

You are free to do so. However, we explicitly rejected alignment from the feature in order to keep it simple and avoid any dependence on the implementation of string.Format or potential culture dependence.

**Comment by ViIvanov** [comment]
Thank you for the answer,
> …keep it simple and avoid any dependence on the implementation of string.Format or potential culture dependence

is enough for me for understanding.

**Comment by IanKemp** [comment]
What happened to this feature? https://github.com/dotnet/roslyn/blob/main/docs/Language%20Feature%20Status.md#c-100 states it was merged into 16.9 yet VS 2019 is currently on 16.11 without this feature - is this coming to VS 2019 ever? Or will it (and C# 10 as a whole) be a VS 2022-only feature?

**Comment by jnm2** [comment]
@IanKemp This feature is working for me in 16.11.2 at the moment. Remember, you have to specify `<LangVersion>preview</LangVersion>` until November when C# 10 is released.

**Comment by IanKemp** [comment]
> @IanKemp This feature is working for me in 16.11.2 at the moment. Remember, you have to specify `<LangVersion>preview</LangVersion>` until November when C# 10 is released.

Thanks for the reminder, for some reason I had it in my head that langver=latest would do it. I also don't see any mention of this feature in the release notes for 16.9.x/16.10.x/16.11.x, which honestly doesn't surprise me because this whole C# language version vs IDE version vs .NET Runtime version matrix is getting rather horribly complicated. I'm hoping you guys are going to perform something of a reset in this regard when you do the big bang drop of C# 10 + .NET 6 + VS 2022 in November.

**Comment by CyrusNajmabadi** [comment]
When c#10 drops, 'latest' will work. Until then, you need 'preview'. This seems same and sensible. 

> is getting rather horribly complicated

I disagree, and I'm an ideal world these can be decoupled.

**Comment by jerviscui** [comment]
Will it solve the interpolated string boxing problem?

eg:

```csharp
int a = 1;
return $"string {a}";
```

**Comment by ufcpp** [comment]
@jerviscui 
see https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/improved-interpolated-strings.md

FYI, interpolation with non-string could not be constant because of formatter dependency.

```cs
Thread.CurrentThread.CurrentCulture = new System.Globalization.CultureInfo("en-us");
Console.WriteLine($"{1.2}"); // 1.2

Thread.CurrentThread.CurrentCulture = new System.Globalization.CultureInfo("fr-fr");
Console.WriteLine($"{1.2}"); // 1,2
```

**Comment by jerviscui** [comment]
@ufcpp  Thank you!
It's means must still use `ToString()` method, to avoid boxing `a`.

```csharp
int a = 1;
return $"string {a.ToString()}";
```

**Comment by drewnoakes** [comment]
Does the formatting of integers change across cultures?

**Comment by jrmoreno1** [comment]
I was surprised to discover that `$”{enum.T}”` wasn’t considered constant…although of course `$”{nameof(enum.T)}”` is
